### ISSUE TITLE
Need cupyx.cusparse.cg()

### ISSUE THREAD

**Issue by adsharma** [cupy/cupy#8837]
### Description

There is a PCG solver sitting in Nvidia's cuda samples repo [here](https://github.com/NVIDIA/CUDALibrarySamples/blob/master/cuSPARSE/cg/cg_example.c#L163-L338).

I would like it to be wrapped as a python function and available via cupy.

More context [here](https://github.com/rapidsai/cugraph/issues/4793)

**Comment by adsharma** [comment]
[This example](https://github.com/cupy/cupy/blob/main/examples/stream/cusparse.py) suggests that there is a `cupyx.cusparse` module. But I couldn't find it.

**Comment by kmaehashi** [comment]
`cupyx.cusparse` is a thin wrapper for cuSPARSE. Basically CuPy's policy is to host NumPy/SciPy-compatible APIs, so specific algorithms like PCG solvers are considered out of scope. I think `cupy.cuda.cusparse.*` APIs can be used to port the example C++ code to CuPy.

**Comment by adsharma** [comment]
Thank you. I was confused by not finding cusparse when I did `dir(cupyx)` or `dir(cupy)`. Looks like you import it dynamically and this is the supported API surface:

https://github.com/cupy/cupy/blob/main/cupy_backends/cuda/cupy_cusparse.h

**Comment by adsharma** [comment]
[This](https://github.com/adsharma/nx-spectral-ordering/blob/main/spectral.py#L56-L77) is the kernel I want to create.

Get the following error:

```
TypeError: <class 'cupyx.scipy.sparse._csr.csr_matrix'> is not supported for RawKernel
```

Is there an example use of  `cupy.cuda.cusparse.*`  I can look at to create a sparse matrix that's compatible?

**Comment by adsharma** [comment]
> algorithms like PCG solvers are considered out of scope

A more narrow question to consider: Is supporting `scipy.sparse.csr_matrix` in `jit.rawkernel` considered in scope?

**Comment by kmaehashi** [comment]
> Thank you. I was confused by not finding cusparse when I did `dir(cupyx)` or `dir(cupy)`. Looks like you import it dynamically and this is the supported API surface:
> 
> https://github.com/cupy/cupy/blob/main/cupy_backends/cuda/cupy_cusparse.h

Yes. Python APIs calling the above functions are defined here: https://github.com/cupy/cupy/blob/e871509d03acbf057b4ba36729be19f8198c3c93/cupy_backends/cuda/libs/cusparse.pyx#L1568

> [This](https://github.com/adsharma/nx-spectral-ordering/blob/main/spectral.py#L56-L77) is the kernel I want to create.

[CuPy JIT](https://docs.cupy.dev/en/latest/user_guide/kernel.html#jit-kernel-definition) generates a single CUDA kernel, so you can't call Python functions within JIT-decorated functions.

> A more narrow question to consider: Is supporting `scipy.sparse.csr_matrix` in `jit.rawkernel` considered in scope?

AFAIK cuSPARSE does not provide device APIs, so this does not make sense.

**Comment by adsharma** [comment]
> [CuPy JIT](https://docs.cupy.dev/en/latest/user_guide/kernel.html#jit-kernel-definition) generates a single CUDA kernel, so you can't call Python functions within JIT-decorated functions.

I looked at `_jit_solve()` closely. It doesn't seem to be calling any generic python functions. Only functions like `cp.dot()` `cp.linalg.norm()` etc. Those should be fusable into a single kernel?

**Comment by kmaehashi** [comment]
I'm closing this one as it's inactive for a while. Feel free to request reopening if necessary.

**Comment by kmaehashi** [comment]
> I looked at `_jit_solve()` closely. It doesn't seem to be calling any generic python functions. Only functions like `cp.dot()` `cp.linalg.norm()` etc. Those should be fusable into a single kernel?

CuPy functions cannot be used inside the JIT. See https://docs.cupy.dev/en/stable/user_guide/kernel.html#jit-kernel-definition.
